source/common/common/hash.h

```diff
@@ -58,7 +61,7 @@ class MurmurHash {
 private:
   static inline uint64_t unalignedLoad(const char* p) {
     uint64_t result;
-    memcpy(&result, p, sizeof(result));
+    safeMemcpy(&result, reinterpret_cast<const uint64_t*>(p));
```

Comment:
Also I believe this `reinterpret_cast` itself is an undefined behavior if p is not aligned. This applies to a couple of casts below.

Reply:
Thanks for the hint 🙂 I will look into this, because in order to apply the safeMemcpy many pointers where reinterpreted so as to count with the size info. Later, I will review if there are places were it would be better to leave ```//NOLINT(safe-memcpy)```

Reply:
I think Lizan raises an interesting point. There's some discussion on https://stackoverflow.com/questions/32589328/unaligned-access-through-reinterpret-cast .

Intuitively I feel like using reinterpret_cast on an unaligned pointer and then using instructions that depend on the alignment is undefined behavior. However, we are only using the result of reinterpret_cast for a static_assert, but then we are just passing the pointer to memcpy. that seems like it should be OK.

However, I think the point of this change is to make it easier to review and gain confidence in code that calls memcpy, and making the code look more suspicious due to UB is probably not the right direction.

Maybe we it would be better to have an alternative safeMemcpy interface for cases where we want to cast the src or dest for any reason. However, after thinking about it for 30 seconds, I couldn't think how any such alternatives would be better in any dimension than just using memcpy directly and declaring NOLINT as you have done in a number of cases. The `reinterpret_cast` is basically saying "trust me!", but it doesn't do it in any more convincing way than using NOLINT, as far as I can tell.

Reply:
I disagree on:

> I couldn't think how any such alternatives would be better in any dimension than just using memcpy directly and declaring NOLINT as you have done in a number of cases

The point of `safeMemcpy` is you don't run the risk of making mistakes around type sizes. It's true if we're casting we could cast to the wrong type, but at least we're expressing a type, which a plain `memcpy` does not do.

Reply:
Cool; I'm not asserting that a better alternative can't be found: just claiming that after thinking about this briefly I didn't find one.

In some of the cases we have an accurate src type and are copying into a char* buffer. In other cases it's the opposite. We could have specialized variants of safeMemcpy for either of those. It's worth a try.

Reply:
Variants make sense, e.g. `safeMemcpySrc`, `safeMemcpyDst`, as long as we are vigilant to make sure they aren't overused.

Reply:
OK, so we'd have something like:
```
uint64_t out;
uint8_t* in;
safeMemcpySrc(&out, in);
```
which would derive the number of bytes to transfer from sizeof(out), and would assume that 'in' has enough bytes. There would nothing we could static_assert to add confidence though, right? And we'd also have:
```
uint8_t* out;
uint64_t in;
safeMemcpyDst(out,. &in);
```
which would take the number of bytes from `in` and we'd have to assume that `out` was big enough.

I think this is OK as long as from a review perspective, we treat those two new forms with the same skepticism that we should if someone annotates `NOLINT` on their memcpy.

Reply:
I agree with what have been said and will implement this. However, I consider that some degree of safety could be atteined by requiring the size of either the src or dst as a parameter of the specialized function, in order to assert this against them, for instance `safeMemcpy(T1* dst, T2* src, size_t size)`. In the previous example, it would be something like:
1.
```
uint64_t out;
uint8_t* in;
safeMemcpySrc(&out, in, 8);
```
2.
```
uint8_t* out;
uint64_t in;
safeMemcpyDst(out, &in, 8);
```

I don't think this is falling back to using a plain memcpy because you will have either `static_assert(size == sizeof(T2))` or `static_assert(sizeof(T1) == size)` to check against.

Reply:
I think whether you write:
```
safeMemcpySrc(&out, in, 8);
```
or
```
safeMemcpySrc(&out, in);
```
the thing that we still need reviewers to do is to hand validate that `in` points to 8 bytes. So it's in that respect not really different, from a reviewability/security perspective, from
```
memcpy(&out, in, 8); // NOLINT(...)
```

The advantage of `safeMemcpySrc(&out, in);` is that it's more terse, because the third arg always must be `sizeof(out)`, and the value of that is lexically obvious (at least on a given platform).